File: app/src/main/java/com/greenwald/aaron/ridetracker/MapFragment.java
package com.greenwald.aaron.ridetracker;

//https://gist.github.com/joshdholtz/4522551

import android.Manifest;
import android.app.Fragment;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.MapsInitializer;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.JointType;
import com.google.android.gms.maps.model.PolylineOptions;
import com.greenwald.aaron.ridetracker.model.TrackPoint;

public class MapFragment extends Fragment {

    MapView mapView;
    GoogleMap map;
    private IntentFilter intentFilter = new IntentFilter();
    PolylineOptions polylineOptions = new PolylineOptions();


    @Override
    public void onResume() {
        super.onResume();
        mapView.onResume();
        getContext().registerReceiver(receiver, intentFilter);
    }

    private BroadcastReceiver receiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (intent.getAction().equals("LOCATION_CHANGED")) {
                String point = intent.getStringExtra("point");
                TrackPoint location = TrackPoint.fromString(point);
                addPointToMap(location);
                focusOnLocation(location);
            }
        }
    };

    private void addPointToMap(TrackPoint location) {
        polylineOptions.add(location.getLatLng());
        map.addPolyline(polylineOptions);
    }

    private void focusOnLocation(TrackPoint location) {
        CameraPosition camPos = new CameraPosition.Builder()
                .target(location.getLatLng())
                .zoom(15)
                .build();
        map.animateCamera(CameraUpdateFactory.newCameraPosition(camPos));
    }

    @Override
    public void onPause() {
        getContext().unregisterReceiver(receiver);
        super.onPause();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, final ViewGroup container, Bundle savedInstanceState) {
        polylineOptions.color(Color.RED);
        polylineOptions.jointType(JointType.ROUND);
        intentFilter.addAction("LOCATION_CHANGED");

        View view = inflater.inflate(R.layout.fragment_map, container, false);
        mapView = view.findViewById(R.id.mapView);
        mapView.onCreate(savedInstanceState);

        mapView.getMapAsync(new OnMapReadyCallback() {
            @Override
            public void onMapReady(GoogleMap googleMap) {
                map = googleMap;
                if (ActivityCompat.checkSelfPermission(getContext(),
                        Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                        && ActivityCompat.checkSelfPermission(getContext(),
                        Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                    return;
                }
                map.setMyLocationEnabled(true);
                map.getUiSettings().setAllGesturesEnabled(true);
                map.getUiSettings().setMyLocationButtonEnabled(true);
            }
        });


        MapsInitializer.initialize(this.getActivity());
        return view;
    }



    @Override
    public void onDestroy() {
        super.onDestroy();
        mapView.onDestroy();
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        mapView.onLowMemory();
    }

}

File: app/src/main/java/com/greenwald/aaron/ridetracker/MapFragment.kt
package com.greenwald.aaron.ridetracker

//https://gist.github.com/joshdholtz/4522551

import android.Manifest
import android.app.Activity
import android.app.AlertDialog
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.AsyncTask
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.MapsInitializer
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.model.JointType
import com.google.android.gms.maps.model.LatLngBounds
import com.google.android.gms.maps.model.PolylineOptions
import com.greenwald.aaron.ridetracker.model.SegmentId
import com.greenwald.aaron.ridetracker.model.SegmentPoint


class MapFragment : Fragment() {

    private lateinit var mapView: MapView
    private lateinit var map: GoogleMap
    private val intentFilter = IntentFilter()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        intentFilter.addAction("LOCATION_CHANGED")

        val view = inflater.inflate(R.layout.fragment_map, container, false)
        mapView = view.findViewById(R.id.mapView)
        mapView.onCreate(savedInstanceState)

        updateData(activity!!)

        return view
    }

    private fun updateData(activity: Activity) {
        AsyncTask.execute {
            val ds = DataStore(activity.applicationContext)
            val trip = ds.getTripWithDetails(arguments!!.getLong("tripId"))

            val segmentPolylines: List<Pair<SegmentId, PolylineOptions>> = trip.segments.map { segment ->
                val locations = segment.segmentPoints.map(SegmentPoint::latLng)

                val speedFactor = segment.maxSpeed.value / trip.maxSpeed.value
                val polyline = PolylineOptions()
                val colorCode = colorForSpeedFactor(speedFactor)
                polyline.color(Color.parseColor(colorCode))
                polyline.jointType(JointType.ROUND)
                polyline.clickable(true)
                polyline.addAll(locations)
                Pair(segment.id, polyline)
            }

            activity.runOnUiThread {
                mapView.getMapAsync(OnMapReadyCallback { googleMap ->
                    map = googleMap
                    if (ActivityCompat.checkSelfPermission(context!!,
                                    Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(context!!,
                                    Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                        return@OnMapReadyCallback
                    }
                    map.isMyLocationEnabled = true
                    map.uiSettings.setAllGesturesEnabled(true)
                    map.uiSettings.isZoomControlsEnabled = true
                    map.uiSettings.isMyLocationButtonEnabled = true

                    segmentPolylines.forEach { segment ->
                        map.addPolyline(segment.second).tag = segment.first
                    }

                    map.setOnPolylineClickListener { polyline ->
                        val segment = ds.getSegment(polyline.tag as SegmentId).toString()
                        showSegmentDialog(segment)
                    }

                    val builder = LatLngBounds.builder()
                    val locations = trip.getAllPoints()
                    if (locations.isNotEmpty()) {
                        locations.forEach { latLng -> builder.include(latLng) }
                        map.moveCamera(CameraUpdateFactory.newLatLngBounds(builder.build(), 200))
                    }
                })

                MapsInitializer.initialize(this.activity!!)
            }


        }
    }

    override fun onAttach(activity: Activity?) {
        super.onAttach(activity)
        (activity as TripActivity).onDataUpdated { tripActivity -> updateData(tripActivity) }
    }

    private fun colorForSpeedFactor(speedFactor: Double): String {
        val red = (speedFactor * 255).toInt().toString(16).padStart(2, '0')
        val green = ((1 - speedFactor) * 255).toInt().toString(16).padStart(2, '0')
        return """#$red${green}00"""
    }

    private fun showSegmentDialog(segment: String) {
        val alertDialog = AlertDialog.Builder(context).create()
        alertDialog.setTitle("Trip Segment")
        alertDialog.setMessage(segment)
        alertDialog.setButton(AlertDialog.BUTTON_NEUTRAL, "OK"
        ) { dialog, _ -> dialog.dismiss() }
        alertDialog.show()
    }


    override fun onDestroy() {
        super.onDestroy()
        mapView.onDestroy()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        mapView.onLowMemory()
    }

    override fun onResume() {
        super.onResume()
        mapView.onResume()
    }

    override fun onPause() {
        super.onPause()
        mapView.onPause()
    }

}